timeboost-sequencer/src/decrypt.rs

```diff
@@ -368,110 +412,111 @@ impl Default for DkgState {
+    state: WorkerState,
 
     /// Number of rounds to retain.
     retain: usize,
 
-    /// Tracker for Dkg bundles received through candidate lists.
+    /// Tracker for DKG bundles received through candidate lists.
     #[builder(default)]
     dkg_tracker: BTreeMap<CommitteeId, DkgAccumulator>,
 
-    /// cache of decrypted shares (keyed by round), each entry value is a nested vector: an
-    /// ordered list of per-ciphertext decryption shares. the order is derived from the
-    /// ciphertext payload from the inclusion list `self.incls` of the same round
-    ///
-    /// note: Option<DecShare> uses None to indicate a failed to decrypt ciphertext
+    /// Tracker for resharing bundles received through candidate lists.
+    #[builder(default)]
+    resharing_tracker: BTreeMap<CommitteeId, ResharingAccumulator>,
+
+    /// Set of committee ids where the accumulation of DkgBundles is completed.
     #[builder(default)]
-    dec_shares: DecSharesCache,
+    dkg_completed: BTreeSet<CommitteeId>,
 
-    /// Acknowledgement of the set of peers whose decryption share for a round has been received
-    /// Useful to prevent DOS or DecShareBatch flooding by malicious peers
+    /// Cache of decryption shares received from remote nodes or produced from local ciphertexts.
+    #[builder(default)]
+    dec_shares: DecShareCache,
+
+    /// Map of received decryption shares for each round.
+    /// Useful to prevent DOS or DecShareBatch flooding by malicious peers.
     #[builder(default)]
     acks: BTreeMap<RoundNumber, HashSet<PublicKey>>,
 
-    /// cache of encrypted inclusion list waiting to be hatched using `dec_shares`
+    /// Inclusion lists (possibly encrypted) waiting to be hatched.
     #[builder(default)]
-    incls: BTreeMap<RoundNumber, InclusionList>,
+    incls: BTreeMap<RoundNumber, (InclusionList, bool)>,
 
-    /// pending inclusion lists not yet decrypted due to missing decryption key.
+    /// Pending inclusion lists not yet decrypted due to missing decryption key.
     #[builder(default)]
     pending: BTreeMap<RoundNumber, InclusionList>,
 
     /// The local clock, driven by round number.
     #[builder(default = RoundNumber::genesis())]
     clock: RoundNumber,
 
-    /// the latest rounds whose ciphertexts are hatched
+    /// The last round for which ciphertexts have hatched.
     #[builder(default = RoundNumber::genesis())]
     last_hatched_round: RoundNumber,
 }
 
 impl Worker {
-    // entry point of `worker` thread, it runs in a loop until shutdown or out of channel capacity.
     pub async fn go(mut self) -> EndOfPlay {
         let node = self.label;
 
-        // always try to catchup first, if other nodes haven't finished, they simply won't respond
-        if let Err(e) = self.dkg_catchup().await {
-            debug!("err during dkg_catchup: {:?}", e);
-            return EndOfPlay::NetworkDown;
+        if !matches!(self.state, WorkerState::AwaitingHandover(_)) {
```

Comment:
we should check the init state, only two valid start state: `AwaitingHandover` and `DkgPending`, we can do a `debug_assert` probably.